File: backend/app/services/formula/engine.py
"""公式引擎:词法分析 + 递归下降解析 + 白名单求值。

纯函数实现,不依赖 DB,方便单测与前端镜像对照。
表达式存储时不带前导 ``=``;变量引用形如 ``{变量名}``。
"""

from __future__ import annotations

import re
from collections import namedtuple
from decimal import Decimal, ROUND_HALF_UP
from typing import Any


class FormulaError(Exception):
    """公式错误(词法/语法/求值)。"""


# ----------------------------- AST 节点 -----------------------------

class Num:
    __slots__ = ("value",)

    def __init__(self, value: float) -> None:
        self.value = value


class Str:
    __slots__ = ("value",)

    def __init__(self, value: str) -> None:
        self.value = value


class Var:
    __slots__ = ("name",)

    def __init__(self, name: str) -> None:
        self.name = name


class UnaryMinus:
    __slots__ = ("operand",)

    def __init__(self, operand: Any) -> None:
        self.operand = operand


class BinOp:
    __slots__ = ("op", "left", "right")

    def __init__(self, op: str, left: Any, right: Any) -> None:
        self.op = op
        self.left = left
        self.right = right


class Compare:
    __slots__ = ("op", "left", "right")

    def __init__(self, op: str, left: Any, right: Any) -> None:
        self.op = op
        self.left = left
        self.right = right


class FuncCall:
    __slots__ = ("name", "args")

    def __init__(self, name: str, args: list[Any]) -> None:
        self.name = name
        self.args = args


WHITELIST_FUNCS = {"ROUND", "MAX", "MIN", "IF", "ABS", "LEN"}

Token = namedtuple("Token", ["type", "value"])


# ----------------------------- Tokenizer -----------------------------

_TOKEN_RE = re.compile(
    r"""
      \s+                                   # 空白(跳过,无捕获组)
    | (?P<NUMBER>\d+\.\d+|\.\d+|\d+)        # 数值
    | \{(?P<VAR>[^}]*)\}                    # {变量名}
    | "(?P<STRING>[^"]*)"                   # "字符串"
    | (?P<FUNC>[A-Za-z_][A-Za-z0-9_]*)      # 标识符(函数名)
    | (?P<CMP><=|>=|<>|=|<|>)               # 比较运算符
    | (?P<OP>[+\-*/^])                      # 算术运算符
    | (?P<LPAREN>\()
    | (?P<RPAREN>\))
    | (?P<COMMA>,)
    """,
    re.VERBOSE,
)

_VAR_REF_RE = re.compile(r"\{([^}]*)\}")


def tokenize(expr: str) -> list[Token]:
    tokens: list[Token] = []
    pos = 0
    length = len(expr)
    while pos < length:
        m = _TOKEN_RE.match(expr, pos)
        if m is None or m.end() == pos:
            raise FormulaError(f"词法错误:无法识别的字符 '{expr[pos]}'")
        kind = m.lastgroup
        if kind is None:  # 空白分支
            pos = m.end()
            continue
        raw = m.group(kind)
        if kind == "NUMBER":
            tokens.append(Token("NUMBER", float(raw)))
        elif kind == "VAR":
            tokens.append(Token("VAR", raw))
        elif kind == "STRING":
            tokens.append(Token("STRING", raw))
        elif kind == "FUNC":
            tokens.append(Token("FUNC", raw.upper()))
        elif kind == "CMP":
            tokens.append(Token("CMP", raw))
        elif kind == "OP":
            tokens.append(Token("OP", raw))
        elif kind == "LPAREN":
            tokens.append(Token("LPAREN", "("))
        elif kind == "RPAREN":
            tokens.append(Token("RPAREN", ")"))
        elif kind == "COMMA":
            tokens.append(Token("COMMA", ","))
        pos = m.end()
    tokens.append(Token("EOF", None))
    return tokens


# ----------------------------- Parser -----------------------------

class Parser:
    """递归下降解析器,``^`` 右结合。"""

    def __init__(self, tokens: list[Token]) -> None:
        self.tokens = tokens
        self.pos = 0

    def _peek(self) -> Token:
        return self.tokens[self.pos]

    def _advance(self) -> Token:
        tok = self.tokens[self.pos]
        self.pos += 1
        return tok

    def _expect(self, type_: str) -> Token:
        tok = self._peek()
        if tok.type != type_:
            raise FormulaError(f"语法错误:期望 {type_},得到 {tok.type}")
        return self._advance()

    def parse(self) -> Any:
        node = self._expr()
        if self._peek().type != "EOF":
            raise FormulaError(f"语法错误:多余的 token '{self._peek().value}'")
        return node

    def _expr(self) -> Any:
        return self._comparison()

    def _comparison(self) -> Any:
        node = self._additive()
        while self._peek().type == "CMP":
            op = self._advance().value
            right = self._additive()
            node = Compare(op, node, right)
        return node

    def _additive(self) -> Any:
        node = self._multiplicative()
        while self._peek().type == "OP" and self._peek().value in ("+", "-"):
            op = self._advance().value
            right = self._multiplicative()
            node = BinOp(op, node, right)
        return node

    def _multiplicative(self) -> Any:
        node = self._power()
        while self._peek().type == "OP" and self._peek().value in ("*", "/"):
            op = self._advance().value
            right = self._power()
            node = BinOp(op, node, right)
        return node

    def _power(self) -> Any:
        base = self._unary()
        if self._peek().type == "OP" and self._peek().value == "^":
            self._advance()
            exp = self._power()  # 右结合:递归解析右侧
            return BinOp("^", base, exp)
        return base

    def _unary(self) -> Any:
        if self._peek().type == "OP" and self._peek().value == "-":
            self._advance()
            return UnaryMinus(self._unary())
        return self._primary()

    def _primary(self) -> Any:
        tok = self._peek()
        if tok.type == "NUMBER":
            self._advance()
            return Num(tok.value)
        if tok.type == "STRING":
            self._advance()
            return Str(tok.value)
        if tok.type == "VAR":
            self._advance()
            return Var(tok.value)
        if tok.type == "FUNC":
            name = self._advance().value
            self._expect("LPAREN")
            args: list[Any] = []
            if self._peek().type != "RPAREN":
                args.append(self._expr())
                while self._peek().type == "COMMA":
                    self._advance()
                    args.append(self._expr())
            self._expect("RPAREN")
            return FuncCall(name, args)
        if tok.type == "LPAREN":
            self._advance()
            node = self._expr()
            self._expect("RPAREN")
            return node
        raise FormulaError(f"语法错误:意外的 token '{tok.value}'")


# ----------------------------- 求值 -----------------------------

def _fmt_num(x: Any) -> str:
    """数值格式化为字面量文本:整数值不带 ``.0``。"""
    if isinstance(x, bool):
        return "1" if x else "0"
    if isinstance(x, float):
        if x.is_integer():
            return str(int(x))
        return repr(x)
    if isinstance(x, int):
        return str(x)
    return str(x)


def _round(x: Any, n: Any) -> float:
    try:
        ndig = int(n)
    except (TypeError, ValueError) as exc:
        raise FormulaError(f"ROUND 的位数参数必须为整数: {exc}") from exc
    try:
        quant = Decimal(1).scaleb(-ndig)
        return float(Decimal(str(x)).quantize(quant, rounding=ROUND_HALF_UP))
    except (ValueError, ArithmeticError) as exc:
        raise FormulaError(f"ROUND 求值失败: {exc}") from exc


def _eval_node(node: Any, values: dict[str, float]) -> Any:
    if isinstance(node, Num):
        return node.value
    if isinstance(node, Str):
        return node.value
    if isinstance(node, Var):
        if node.name not in values:
            raise FormulaError(f"未绑定变量: {node.name}")
        return values[node.name]
    if isinstance(node, UnaryMinus):
        return -_eval_node(node.operand, values)
    if isinstance(node, BinOp):
        left = _eval_node(node.left, values)
        right = _eval_node(node.right, values)
        op = node.op
        if op == "+":
            return left + right
        if op == "-":
            return left - right
        if op == "*":
            return left * right
        if op == "/":
            if right == 0:
                raise FormulaError("除零错误")
            return left / right
        if op == "^":
            try:
                return float(left ** right)
            except (ValueError, OverflowError, TypeError, ZeroDivisionError) as exc:
                raise FormulaError(f"幂运算错误: {exc}") from exc
        raise FormulaError(f"未知运算符: {op}")
    if isinstance(node, Compare):
        left = _eval_node(node.left, values)
        right = _eval_node(node.right, values)
        op = node.op
        if op == ">":
            return 1.0 if left > right else 0.0
        if op == "<":
            return 1.0 if left < right else 0.0
        if op == ">=":
            return 1.0 if left >= right else 0.0
        if op == "<=":
            return 1.0 if left <= right else 0.0
        if op == "=":
            return 1.0 if left == right else 0.0
        if op == "<>":
            return 1.0 if left != right else 0.0
        raise FormulaError(f"未知比较运算符: {op}")
    if isinstance(node, FuncCall):
        return _eval_func(node, values)
    raise FormulaError("未知 AST 节点")


def _eval_func(node: FuncCall, values: dict[str, float]) -> Any:
    name = node.name
    args = node.args
    if name not in WHITELIST_FUNCS:
        raise FormulaError(f"未知函数: {name}")
    if name == "IF":
        if len(args) != 3:
            raise FormulaError("IF 需要 3 个参数")
        cond = _eval_node(args[0], values)
        return _eval_node(args[1], values) if cond != 0 else _eval_node(args[2], values)
    if name == "ROUND":
        if len(args) != 2:
            raise FormulaError("ROUND 需要 2 个参数")
        return _round(_eval_node(args[0], values), _eval_node(args[1], values))
    if name == "MAX":
        if len(args) < 1:
            raise FormulaError("MAX 至少需要 1 个参数")
        return max(_eval_node(a, values) for a in args)
    if name == "MIN":
        if len(args) < 1:
            raise FormulaError("MIN 至少需要 1 个参数")
        return min(_eval_node(a, values) for a in args)
    if name == "ABS":
        if len(args) != 1:
            raise FormulaError("ABS 需要 1 个参数")
        return abs(_eval_node(args[0], values))
    if name == "LEN":
        if len(args) != 1:
            raise FormulaError("LEN 需要 1 个参数")
        val = _eval_node(args[0], values)
        if isinstance(val, str):
            return len(val)
        return len(_fmt_num(val))
    raise FormulaError(f"未知函数: {name}")


def _build_substituted(expression: str, values: dict[str, float]) -> str:
    def repl(m: re.Match[str]) -> str:
        name = m.group(1)
        if name in values:
            return _fmt_num(values[name])
        return m.group(0)

    return _VAR_REF_RE.sub(repl, expression)


def evaluate(expression: str, values: dict[str, float]) -> dict[str, Any]:
    """求值,返回 ``{value, substituted_expression, db_refs}``。

    ``db_refs`` 在本接口始终为空列表(来源信息由解析流水线后续填充)。
    """
    if expression is None or not expression.strip():
        raise FormulaError("表达式为空")
    tokens = tokenize(expression)
    ast = Parser(tokens).parse()
    value = _eval_node(ast, values)
    substituted = _build_substituted(expression, values)
    return {"value": value, "substituted_expression": substituted, "db_refs": []}


def extract_var_refs(expression: str) -> list[str]:
    """提取表达式中 ``{var}`` 引用的变量名,按首次出现顺序去重。"""
    seen: set[str] = set()
    ordered: list[str] = []
    for name in _VAR_REF_RE.findall(expression or ""):
        if name not in seen:
            seen.add(name)
            ordered.append(name)
    return ordered


__all__ = [
    "FormulaError",
    "evaluate",
    "extract_var_refs",
    "tokenize",
    "Parser",
    "WHITELIST_FUNCS",
]

File: frontend/src/lib/formula/engine.ts
/**
 * 公式引擎(前端 TS 镜像,与后端 Python 严格一致)。
 *
 * 文法:
 *   expr        -> comparison
 *   comparison  -> additive ( (> | < | >= | <= | = | <>) additive )*
 *   additive    -> multiplicative ( (+ | -) multiplicative )*
 *   multiplicative -> power ( (* | /) power )*
 *   power       -> unary ( ^ power )?     // 右结合
 *   unary       -> (- | +)? primary
 *   primary     -> number | {var} | string | func(args) | ( expr )
 *
 * 变量引用 {name} 替换为 values[name];缺失抛 Error("未绑定变量: name")。
 * 比较运算符产生 1/0。^ 右结合。
 * 函数白名单(大小写不敏感):ROUND MAX MIN IF ABS LEN
 */

export interface EvaluateResult {
  value: number
  substitutedExpression: string
  dbRefs: unknown[]
}

type Value = number | string

type TokenType =
  | 'NUMBER'
  | 'VAR'
  | 'STRING'
  | 'IDENT'
  | 'PLUS'
  | 'MINUS'
  | 'STAR'
  | 'SLASH'
  | 'CARET'
  | 'LPAREN'
  | 'RPAREN'
  | 'COMMA'
  | 'GT'
  | 'LT'
  | 'GE'
  | 'LE'
  | 'EQ'
  | 'NE'
  | 'EOF'

interface Token {
  type: TokenType
  /** 原始文本(变量名/标识符/字符串内容) */
  value: string
  num?: number
  pos: number
}

function isDigit(c: string): boolean {
  return c >= '0' && c <= '9'
}

function isIdentStart(c: string): boolean {
  return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c === '_'
}

function isIdentPart(c: string): boolean {
  return isIdentStart(c) || isDigit(c)
}

/** 格式化数值:整数去 .0 */
function formatNum(v: number): string {
  if (Number.isInteger(v)) return String(v)
  return String(v)
}

function tokenize(expr: string): Token[] {
  const tokens: Token[] = []
  let i = 0
  const n = expr.length
  while (i < n) {
    const c = expr[i]
    // 空白
    if (c === ' ' || c === '\t' || c === '\n' || c === '\r') {
      i++
      continue
    }
    const pos = i
    // 数字
    if (isDigit(c) || (c === '.' && isDigit(expr[i + 1] ?? ''))) {
      let j = i
      let dotSeen = false
      while (j < n && (isDigit(expr[j]) || (expr[j] === '.' && !dotSeen))) {
        if (expr[j] === '.') dotSeen = true
        j++
      }
      const text = expr.slice(i, j)
      tokens.push({ type: 'NUMBER', value: text, num: parseFloat(text), pos })
      i = j
      continue
    }
    // 变量引用 {name}(支持中文)
    if (c === '{') {
      let j = i + 1
      while (j < n && expr[j] !== '}') j++
      if (j >= n) throw new Error('词法错误: 未闭合的变量引用 {')
      const name = expr.slice(i + 1, j).trim()
      tokens.push({ type: 'VAR', value: name, pos })
      i = j + 1
      continue
    }
    // 字符串字面量
    if (c === '"') {
      let j = i + 1
      let s = ''
      while (j < n && expr[j] !== '"') {
        s += expr[j]
        j++
      }
      if (j >= n) throw new Error('词法错误: 未闭合的字符串 "')
      tokens.push({ type: 'STRING', value: s, pos })
      i = j + 1
      continue
    }
    // 标识符(函数名)
    if (isIdentStart(c)) {
      let j = i
      while (j < n && isIdentPart(expr[j])) j++
      tokens.push({ type: 'IDENT', value: expr.slice(i, j), pos })
      i = j
      continue
    }
    // 双字符运算符
    const two = expr.slice(i, i + 2)
    if (two === '>=') {
      tokens.push({ type: 'GE', value: '>=', pos })
      i += 2
      continue
    }
    if (two === '<=') {
      tokens.push({ type: 'LE', value: '<=', pos })
      i += 2
      continue
    }
    if (two === '<>') {
      tokens.push({ type: 'NE', value: '<>', pos })
      i += 2
      continue
    }
    // 单字符运算符
    switch (c) {
      case '+':
        tokens.push({ type: 'PLUS', value: '+', pos })
        break
      case '-':
        tokens.push({ type: 'MINUS', value: '-', pos })
        break
      case '*':
        tokens.push({ type: 'STAR', value: '*', pos })
        break
      case '/':
        tokens.push({ type: 'SLASH', value: '/', pos })
        break
      case '^':
        tokens.push({ type: 'CARET', value: '^', pos })
        break
      case '(':
        tokens.push({ type: 'LPAREN', value: '(', pos })
        break
      case ')':
        tokens.push({ type: 'RPAREN', value: ')', pos })
        break
      case ',':
        tokens.push({ type: 'COMMA', value: ',', pos })
        break
      case '>':
        tokens.push({ type: 'GT', value: '>', pos })
        break
      case '<':
        tokens.push({ type: 'LT', value: '<', pos })
        break
      case '=':
        tokens.push({ type: 'EQ', value: '=', pos })
        break
      default:
        throw new Error(`词法错误: 无法识别的字符 "${c}"`)
    }
    i++
  }
  tokens.push({ type: 'EOF', value: '', pos: n })
  return tokens
}

class Parser {
  private tokens: Token[]
  private pos = 0
  private values: Record<string, number>

  constructor(tokens: Token[], values: Record<string, number>) {
    this.tokens = tokens
    this.values = values
  }

  private peek(): Token {
    return this.tokens[this.pos]
  }

  private next(): Token {
    return this.tokens[this.pos++]
  }

  private expect(type: TokenType): Token {
    const t = this.tokens[this.pos]
    if (t.type !== type) {
      throw new Error(`语法错误: 期望 ${type} 但得到 ${t.type}`)
    }
    return this.tokens[this.pos++]
  }

  parse(): Value {
    const v = this.parseExpr()
    if (this.peek().type !== 'EOF') {
      throw new Error('语法错误: 表达式后存在多余内容')
    }
    return v
  }

  private parseExpr(): Value {
    return this.parseComparison()
  }

  private parseComparison(): Value {
    let left = this.parseAdditive()
    for (;;) {
      const t = this.peek().type
      if (
        t !== 'GT' &&
        t !== 'LT' &&
        t !== 'GE' &&
        t !== 'LE' &&
        t !== 'EQ' &&
        t !== 'NE'
      ) {
        break
      }
      this.next()
      const right = this.parseAdditive()
      const l = this.toNum(left)
      const r = this.toNum(right)
      let res: number
      switch (t) {
        case 'GT':
          res = l > r ? 1 : 0
          break
        case 'LT':
          res = l < r ? 1 : 0
          break
        case 'GE':
          res = l >= r ? 1 : 0
          break
        case 'LE':
          res = l <= r ? 1 : 0
          break
        case 'EQ':
          res = l === r ? 1 : 0
          break
        case 'NE':
          res = l !== r ? 1 : 0
          break
        default:
          res = 0
      }
      left = res
    }
    return left
  }

  private parseAdditive(): Value {
    let left = this.parseMultiplicative()
    for (;;) {
      const t = this.peek().type
      if (t !== 'PLUS' && t !== 'MINUS') break
      this.next()
      const right = this.parseMultiplicative()
      const l = this.toNum(left)
      const r = this.toNum(right)
      left = t === 'PLUS' ? l + r : l - r
    }
    return left
  }

  private parseMultiplicative(): Value {
    let left = this.parsePower()
    for (;;) {
      const t = this.peek().type
      if (t !== 'STAR' && t !== 'SLASH') break
      this.next()
      const right = this.parsePower()
      const l = this.toNum(left)
      const r = this.toNum(right)
      if (t === 'STAR') {
        left = l * r
      } else {
        if (r === 0) throw new Error('除零错误')
        left = l / r
      }
    }
    return left
  }

  private parsePower(): Value {
    const base = this.parseUnary()
    if (this.peek().type === 'CARET') {
      this.next()
      // 右结合:递归调用 parsePower
      const exp = this.parsePower()
      const b = this.toNum(base)
      const e = this.toNum(exp)
      return Math.pow(b, e)
    }
    return base
  }

  private parseUnary(): Value {
    const t = this.peek().type
    if (t === 'MINUS') {
      this.next()
      return -this.toNum(this.parseUnary())
    }
    if (t === 'PLUS') {
      this.next()
      return this.parseUnary()
    }
    return this.parsePrimary()
  }

  private parsePrimary(): Value {
    const t = this.peek()
    switch (t.type) {
      case 'NUMBER':
        this.next()
        return t.num!
      case 'VAR': {
        this.next()
        const name = t.value
        if (!(name in this.values)) {
          throw new Error(`未绑定变量: ${name}`)
        }
        return this.values[name]
      }
      case 'STRING':
        this.next()
        return t.value
      case 'LPAREN': {
        this.next()
        const v = this.parseExpr()
        this.expect('RPAREN')
        return v
      }
      case 'IDENT':
        return this.parseFunction()
      default:
        throw new Error(`语法错误: 意外的 token ${t.type}`)
    }
  }

  private parseFunction(): Value {
    const nameTok = this.next()
    const origName = nameTok.value
    const name = origName.toUpperCase()
    this.expect('LPAREN')
    const args: Value[] = []
    if (this.peek().type !== 'RPAREN') {
      args.push(this.parseExpr())
      while (this.peek().type === 'COMMA') {
        this.next()
        args.push(this.parseExpr())
      }
    }
    this.expect('RPAREN')
    return this.callFunction(name, origName, args)
  }

  private callFunction(name: string, origName: string, args: Value[]): Value {
    switch (name) {
      case 'ROUND': {
        if (args.length !== 2) {
          throw new Error('参数数量错误: ROUND 需要 2 个参数')
        }
        const x = this.toNum(args[0])
        const n = this.toNum(args[1])
        const factor = Math.pow(10, n)
        return Math.round(x * factor) / factor
      }
      case 'MAX': {
        if (args.length < 1) {
          throw new Error('参数数量错误: MAX 至少需要 1 个参数')
        }
        return Math.max(...args.map((a) => this.toNum(a)))
      }
      case 'MIN': {
        if (args.length < 1) {
          throw new Error('参数数量错误: MIN 至少需要 1 个参数')
        }
        return Math.min(...args.map((a) => this.toNum(a)))
      }
      case 'IF': {
        if (args.length !== 3) {
          throw new Error('参数数量错误: IF 需要 3 个参数')
        }
        const c = this.toNum(args[0])
        return c !== 0 ? args[1] : args[2]
      }
      case 'ABS': {
        if (args.length !== 1) {
          throw new Error('参数数量错误: ABS 需要 1 个参数')
        }
        return Math.abs(this.toNum(args[0]))
      }
      case 'LEN': {
        if (args.length !== 1) {
          throw new Error('参数数量错误: LEN 需要 1 个参数')
        }
        const a = args[0]
        const s = typeof a === 'string' ? a : formatNum(this.toNum(a))
        return s.length
      }
      default:
        throw new Error(`未知函数: ${origName}`)
    }
  }

  private toNum(v: Value): number {
    if (typeof v === 'number') return v
    throw new Error('类型错误: 期望数值但得到字符串')
  }
}

/**
 * 求值公式表达式。
 * @param expression 公式文本,变量用 {name} 引用
 * @param values 变量名 -> 数值
 * @returns { value, substitutedExpression, dbRefs }
 */
export function evaluate(
  expression: string,
  values: Record<string, number>
): EvaluateResult {
  const tokens = tokenize(expression)
  const parser = new Parser(tokens, values)
  const result = parser.parse()
  if (typeof result !== 'number') {
    throw new Error('表达式结果不是数值')
  }
  // substitutedExpression: 把 {var} 替换成数值(整数去 .0)
  const substitutedExpression = expression.replace(
    /\{([^}]+)\}/g,
    (_m, name: string) => {
      const key = name.trim()
      if (!(key in values)) {
        throw new Error(`未绑定变量: ${key}`)
      }
      return formatNum(values[key])
    }
  )
  return { value: result, substitutedExpression, dbRefs: [] }
}

/**
 * 提取表达式中所有 {var} 引用名,去重保序。
 */
export function extractVarRefs(expression: string): string[] {
  const refs: string[] = []
  const seen = new Set<string>()
  const re = /\{([^}]+)\}/g
  let m: RegExpExecArray | null
  while ((m = re.exec(expression)) !== null) {
    const name = m[1].trim()
    if (!seen.has(name)) {
      seen.add(name)
      refs.push(name)
    }
  }
  return refs
}
